lib/spack/spack/build_environment.py

```diff
@@ -343,10 +343,11 @@ def set_compiler_environment_variables(pkg, env):
         if inject_flags[flag]:
             # variables SPACK_<FLAG> inject flags through wrapper
             var_name = "SPACK_{0}".format(flag.upper())
-            env.set(var_name, " ".join(f for f in inject_flags[flag]))
+            env.set(var_name, " ".join(f[0] for f in inject_flags[flag]))
         if env_flags[flag]:
             # implicit variables
-            env.set(flag.upper(), " ".join(f for f in env_flags[flag]))
+            env.set(flag.upper(), " ".join(f[0] for f in env_flags[flag]))
```

Comment:
I realized this implementation is going to break custom flag handlers.

We either need to be able to iterate over the flags and their propagate values separately or we need to only pass the flags (not the propagate values) to the flag handler.

We could switch from `[(, ), (, )]` to two ordered lists, or we could call the flag handler differently. And then we can remove these changes.

I'm partial to the option where we store two lists, because it allows package writers to access the flags in a more intuitive way. Basically `compiler_flags[name]` would give just the flags, and we add another data structure `compiler_flags._propagation[name]` gives the propagation values. Then we can define `compiler_flags.with_propagation()` to return the tuples the way you've constructed them, for places we need that information.

We can overwrite whatever methods we need to be able to extend this interface to other ways we interact with the `FlagMap` object.